src/NLog/LoggerImpl.cs

```diff
@@ -114,21 +114,54 @@ private static int FindCallingMethodOnStackTrace([NotNull] StackTrace stackTrace
             var stackFrames = stackTrace.GetFrames();
             if (stackFrames == null)
                 return 0;
-            var intermediate = stackFrames.Select((f, i) => new {index = i, frame = f});
+            var intermediate = stackFrames.Select((f, i) => new StackFrameWithIndex(i, f));
             //find until logger type
-            intermediate = intermediate.SkipWhile(p => !IsLoggerType(p.frame, loggerType));
-            //skip to next
-            intermediate = intermediate.Skip(1);
-            //skip while in "skip" assemlbl
-            intermediate = intermediate.SkipWhile(p => SkipAssembly(p.frame));
+            intermediate = intermediate.SkipWhile(p => !IsLoggerType(p.StackFrame, loggerType));
+
+            intermediate = FilterBySkipAssembly(intermediate);
+            return GetIndexOfNonAsync(intermediate);
+        }
+
+        private static int GetIndexOfNonAsync(IEnumerable<StackFrameWithIndex> intermediate)
+        {
             var last = intermediate.FirstOrDefault();
 
             if (last != null)
             {
-                return last.index;
+#if ASYNC_SUPPORTED
+
+                //movenext and then AsyncTaskMethodBuilder (method start)? this is a generated MoveNext by async.
+                if (last.StackFrame.GetMethod().Name == "MoveNext")
+                {
+                    var next = intermediate.Skip(1).FirstOrDefault();
+                    if (next != null)
+                    {
+                        var declaringType = next.StackFrame.GetMethod().DeclaringType;
+                        if (declaringType != null && declaringType.FullName == "System.Runtime.CompilerServices.AsyncTaskMethodBuilder")
+                        {
+
+                            //async, search futher
+
+                            intermediate = FilterBySkipAssembly(intermediate);
+                            return GetIndexOfNonAsync(intermediate);
+                        }
+                    }
+                }
+#endif
+
+                return last.StackFrameIndex;
             }
             return 0;
+        }
+
+        private static IEnumerable<StackFrameWithIndex> FilterBySkipAssembly(IEnumerable<StackFrameWithIndex> intermediate)
+        {
 
+            //skip to next
+            intermediate = intermediate.Skip(1);
```

Comment:
I'm not sure this line belongs inside this method. The name of the method does not say that it's skipping the next element.